[isoltest] Support FAILURE with reason

erak opened the pull request:
Part of https://github.com/ethereum/solidity/issues/6669.

### Desctiption ###
This PR adds parsing and formatting support for e.g. revert reasons that were returned by the call:
```
contract C {
    function e() public {
        revert("Transaction failed.");
    }
}
// ====
// EVMVersion: >homestead
// ----
// _() -> FAILURE
// e() -> FAILURE, hex"08c379a0", 0x20, 19, "Transaction failed."
```
It also splits up the smoke tests for `isoltest` and includes a small refactoring towards a cleaner and safer formatting of results returned from calls.

chriseth commented:
Sounds good! Don't you need to add additional info to signal that `hex"08c379a0"` is not padded?

erak commented:
The "workflow" I had in mind implies, that you would see the actual result returned by the call once (added with https://github.com/ethereum/solidity/pull/7052). This based on my guess, that you won't pre-compute the signature before writing a test which means that would probably start off with:

`// e() -> FAILURE, hex"00", 0x00, "Transaction failed."`

which prints the actual returned bytes:

```
Obtained result:
// e() -> FAILURE, hex"08c379a0", 0x20, 13, "Transaction failed."

Warning: the call to "f()" returned
[08,c3,79,a0]
[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,20]
[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,13]
[54,72,61,6e,73,61,63,74,69,6f,6e,20,66,61,69,6c,65,64,2e,0,0,0,0,0,0,0,0,0,0,0,0,0]
```
and that should make it clear, that the signature is not padded. Does this makes sense to you?

chriseth commented:
Does this mean that not padding it is a special case for this situation or not? Do we pad `hex"..."` in other situations or not?

erak commented:
@chriseth No padding for `hex"..."` is the default and works with function arguments and with expectations that actually return unpadded bytes (like in this PR). We had no test cases that tested functions returning unpadded results before.

Formatting fails when target type is a (unpadded) hex string or a (right-aligned, padded) hex number and the actual result is left-aligned and padded.

erak commented:
Ah, ok. The wrong formatting of left-aligned, padded bytes for target type hex number will be fixed with https://github.com/ethereum/solidity/pull/7091.

Then, I think we should merge https://github.com/ethereum/solidity/pull/7052 and https://github.com/ethereum/solidity/pull/7091 first.

chriseth commented:
Ok, then it's fine for this PR. Have we talked about requiring `unpadded(...)` in such cases?

chriseth commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -154,40 +163,87 @@ string TestFunctionCall::formatBytesParameters(
-		m_contractABI,
-		_signature
-	);
+		os << BytesUtils::formatBytesRange(_bytes, defaultParameters, _highlight);
```
What happens if the data is not according to the given types above? This may well be the case later.

erak replied:
The format would break. We might want to check if given bytes *can* be formatted before every call to `formatBytesRange`. But I would keep this in mind and implement it at a later point in time.

chriseth replied:
I would prefer something that is working in all cases and just is not nice in some cases instead of it breaking in some cases.

erak replied:
Wait, are you talking about the "special" case of a `FAILURE` or in general? I just checked and got a weird output in case the returned bytes do not match the given encoding:
```
function u() public returns (T[2] memory) {
    return [T(23, 42, "any"), T(555, 666, "any")];
}
// ----
// u() -> 0x20, 0x40, 0xE0, 23, 42, 0x60, 3, hex"ef0042", 555, 666, 0x60, 3
```
results in
```
// u() -> 0x20, 0x40, 0xE0, 23, 42, 0x60, 3, hex"616e79", 0, 9311354880, 0x29A000000, 1610612736, 56716921
```
which breaks the test after updating it. At least you have the actual returned data printed as a hint:
```
  Warning: The call to "u()" returned 
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,20]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,40]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,e0]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,17]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2a]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,60]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3]
  [61,6e,79,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,2b]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,9a]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,60]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3]
  [61,6e,79,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
```

Fixing this might be not in the scope of this PR. Shall we create an issue?

chriseth replied:
Yes, please create an issue.

chriseth commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -154,40 +163,87 @@ string TestFunctionCall::formatBytesParameters(
+			ParameterList defaultParameters;
+			fill_n(
+				back_inserter(defaultParameters),
+				_bytes.size() / 32,
```
Is rounding down the correct behaviour?

chriseth commented on test/libsolidity/semanticTests/smoke_test.sol:
```diff
@@ -84,11 +87,14 @@ contract C {
 // _() -> FAILURE
-// e() ->
+// d() ->
+// e() -> FAILURE, hex"08c379a0", 0x20, 19, "Transaction failed."
```
Can you also add an example with a string longer than 32 bytes?

erak replied:
Supporting strings longer than 32 bytes is a bit more work and out-of-scope, I'd say. Created an issue for it: https://github.com/ethereum/solidity/issues/7197.

leonardoalt commented:
Please rebase

erak committed:
[isoltest] Support FAILURE with reason.

erak commented on test/libsolidity/util/SoltestTypes.h:
```diff
@@ -139,6 +139,7 @@ struct ABIType
 struct FormatInfo
 {
 	bool newline = false;
+	bool hide = false;
```
This could be used to hide parameters from the output e.g.:
```
// e() -> FAILURE, hex"08c379a0", 0x20, 19, "Transaction failed."
```
becomes
```
// e() -> FAILURE, "Transaction failed."
```
Even though we decided to not hide the signature, the tail-pointer and the length (see discussion), I left it in since it could be useful in the future.

Marenz replied:
Maybe add this as comment in the code too? :)

chriseth replied:
I would actually prefer to remove that functionality, if it is off by default.

erak replied:
I agree and removed it.

erak commented:
@Marenz Would like to get a review since this is your feature request ;)

chriseth commented on test/libsolidity/semanticTests/smoke_test.sol:
```diff
@@ -88,12 +91,15 @@ contract C {
     }
 }
+// ====
+// EVMVersion: >homestead
```
Please split up the test.

erak replied:
Done.

Marenz commented on test/libsolidity/util/BytesUtils.cpp:
```diff
@@ -206,21 +206,28 @@ string BytesUtils::formatString(bytes const& _bytes, size_t _cutOff)
 }
 
-string BytesUtils::formatRawBytes(bytes const& _bytes)
+string BytesUtils::formatRawBytes(bytes const& _bytes, string _linePrefix, bool _withSignature)
```
just a thought: Would it make sense to make the parameter a `bytes _bytes` (no  const, no ref) and modify it after the signature instead of creating `tail`?

erak replied:
That's what I had at some point, but passing it by const-ref came out of another review. Also, I had some upcoming changes in mind (formatting weirdly-sized return values) which could make it mandatory to modularize this function.

Marenz commented on test/libsolidity/util/TestFileParserTests.cpp:
```diff
@@ -219,6 +219,23 @@ BOOST_AUTO_TEST_CASE(non_existent_call_revert)
+BOOST_AUTO_TEST_CASE(call_revert_message)
+{
+	char const* source = R"(
+		// f() -> FAILURE, hex"08c379a0", 0x20, 6, "Revert"
```
Very nice, I'm gonna need exactly this use case

Marenz reviewed (dismissed):
A few suggestions but nothing preventing a merge.

chriseth commented:
Please rebase.

erak committed:
[isoltest] Splits smoke test into separate files.

erak commented:
> Ok, then it's fine for this PR. Have we talked about requiring `unpadded(...)` in such cases?

@chriseth What's the difference between `unpadded(0xEF0042)` and `hex"EF0042"`? Would you prefer to also have `unpadded(...)` for padded types like hexadecimals?

erak commented:
Rebased and updated. I took care of all comments except from https://github.com/ethereum/solidity/pull/7095/files#r304959011, which seems to be also out-of-scope.

erak commented:
Fixed a MSVC issue.

chriseth commented:
Not sure how we decidid that in the past, but currently, I would assume everything to be padded unless explicitly specified otherwise. Does not need to be changed in this PR, though.

chriseth commented on test/libsolidity/util/BytesUtils.cpp:
```diff
@@ -202,21 +202,28 @@ string BytesUtils::formatString(bytes const& _bytes, size_t _cutOff)
+	bytes tail{it, _bytes.end()};
+	it = tail.begin();
 
+	for (size_t i = 0; i < tail.size(); i += 32)
```
Doesn't this need `i + 32 <= tail.size()` if you use `it + 32` as end below?

erak replied:
I've refactored the formatting a little bit and it's now all based on lists of `Parameter`. Please see the last commit.

This way we can at least assert that `_bytes.size() == encodingSize(parameters)` in functions that are supposed to format bytes in some way and avoid dealing with such iterator offsets altogether.

chriseth commented on test/libsolidity/util/BytesUtils.cpp:
```diff
@@ -202,21 +202,28 @@ string BytesUtils::formatString(bytes const& _bytes, size_t _cutOff)
+	if (_withSignature)
 	{
-		bytes byteRange{it, it + 32};
+		os << _linePrefix << bytes{it, it + 4} << endl;
```
Do we know if `it + 4` is valid?

erak replied:
Removed. Please see last commit.

erak committed:
[isoltest] Refactors parameter formatting.

erak commented:
> Not sure how we decidid that in the past, but currently, I would assume everything to be padded unless explicitly specified otherwise. Does not need to be changed in this PR, though.

@chriseth Yes, that's the case. The only exception is `hex"ef0042"`, which is not padded. I can't fully remember why we decided to favor `hex"ef0042"` over `unpadded(0xef0042)`. Maybe because `left(...)` and `right(...)` were not there yet.

chriseth commented on test/libsolidity/semanticTests/smoke/alignment.sol:
```diff
@@ -26,3 +26,4 @@ contract D {
 // stateBytes() -> left(0x4200ef)
 // internalStateDecimal() -> 0x20
 // update(bool,uint256,bytes32): false, -23, left(0x2300ef) -> false, -23, left(0x2300ef)
+
```
Why the newlines?
